File: app/src/main/java/com/marpak/gmpolygon/MainActivity.java
package com.marpak.gmpolygon;

import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;

import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.SupportMapFragment;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Polygon;
import com.google.android.gms.maps.model.PolygonOptions;
import com.google.android.gms.maps.model.Polyline;
import com.google.android.gms.maps.model.PolylineOptions;
import com.marpak.gmpolygon.databinding.ActivityMainBinding;
import com.marpak.gmpolygon.enums.ButtonAction;

import java.util.ArrayList;
import java.util.List;

public class MainActivity extends AppCompatActivity implements OnMapReadyCallback {

    private GoogleMap googleMap;
    private GoogleMap.OnMapClickListener mapClickListener;
    private final List<LatLng> polylinePoints = new ArrayList<>();
    private Polyline polyline;
    private Polygon polygon;

    private ActivityMainBinding binding;
    private ButtonPalette buttonPalette;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        // Set the layout file as the content view.
        binding = ActivityMainBinding.inflate(getLayoutInflater());
        View view = binding.getRoot();
        setContentView(view);

        // Get a handle to the fragment and register the callback.
        SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager().findFragmentById(R.id.fragMap);
        if (mapFragment != null) {
            mapFragment.getMapAsync(this);
        }

        buttonPalette = new ButtonPalette(binding.buttonLayout);
    }

    // Get a handle to the GoogleMap object and display marker.
    @Override
    public void onMapReady(@NonNull GoogleMap map) {
        googleMap = map;

        mapClickListener = this::addPoint;

        binding.btnDrawPolygon.setOnClickListener(view -> {
            if (buttonPalette.getCurrentAction() == ButtonAction.CLEAR_DRAWINGS) {
                buttonPalette.setAction(ButtonAction.START_DRAWING_SHAPE);

                // start collecting click points
                polylinePoints.clear();
                googleMap.setOnMapClickListener(mapClickListener);

            } else if (buttonPalette.getCurrentAction() == ButtonAction.START_DRAWING_SHAPE) {
                // stop collecting click points and draw polygon
                if(polylinePoints.size() >=3) {
                    drawPolygon();

                    buttonPalette.setAction(ButtonAction.STOP_DRAWING_SHAPE);

                    polylinePoints.clear();
                    googleMap.setOnMapClickListener(null);
                }
            }
        });

        binding.btnDrawCrossLine.setOnClickListener(view -> {
            if (buttonPalette.getCurrentAction() == ButtonAction.STOP_DRAWING_SHAPE) {
                buttonPalette.setAction(ButtonAction.START_DRAWING_CROSS_LINE);

                // start collecting click points
                polylinePoints.clear();
                googleMap.setOnMapClickListener(mapClickListener);

            } else if (buttonPalette.getCurrentAction() == ButtonAction.START_DRAWING_CROSS_LINE) {
                // stop collecting click points and draw cross line
                if(polylinePoints.size() >=2) {
                    buttonPalette.setAction(ButtonAction.STOP_DRAWING_CROSS_LINE);

                    polylinePoints.clear();
                    googleMap.setOnMapClickListener(null);
                }
            }
        });

        binding.btnDivideCrossLine.setOnClickListener(view -> {
            if (buttonPalette.getCurrentAction() == ButtonAction.STOP_DRAWING_CROSS_LINE) {
                // get user input via an prompt
                InputPromptDialog.showInputDialog(MainActivity.this,
                        getResources().getString(R.string.prompt_dialog_title),
                        getResources().getString(R.string.prompt_dialog_description),
                        userInput -> {
                    // divide the cross line
                    buttonPalette.setAction(ButtonAction.DIVIDE_CROSS_LINE);
                }, () -> {

                });
            }
        });

        binding.btnClearDrawings.setOnClickListener(view -> {
            polygon.remove();
            polyline.remove();

            buttonPalette.setAction(ButtonAction.CLEAR_DRAWINGS);
        });
    }

    private void addPoint(LatLng latLng) {
        polylinePoints.add(latLng);
        if (polyline != null) {
            // remove existing polyline from map and add new polyline with updated points
            polyline.remove();
        }
        polyline = googleMap.addPolyline(new PolylineOptions().addAll(polylinePoints).color(getResources().getColor(R.color.polyline_color)));
    }

    private void drawPolygon() {
        if (polygon != null) {
            // remove existing polygon from map and add new polygon with updated points
            polygon.remove();
        }
        polygon = googleMap.addPolygon(new PolygonOptions().addAll(polylinePoints).strokeColor(getResources().getColor(R.color.polyline_color)).fillColor(getResources().getColor(R.color.polygon_fill_color)));
        polygon.setTag("User Selection");
    }
}

File: app/src/main/java/com/marpak/gmpolygon/MainActivity.kt
package com.marpak.gmpolygon

import android.os.Bundle
import android.view.View
import android.widget.Button
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.GoogleMap.OnMapClickListener
import com.google.android.gms.maps.GoogleMap.OnMarkerDragListener
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.SupportMapFragment
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Marker
import com.google.android.gms.maps.model.MarkerOptions
import com.google.android.gms.maps.model.Polygon
import com.google.android.gms.maps.model.PolygonOptions
import com.google.android.gms.maps.model.Polyline
import com.google.android.gms.maps.model.PolylineOptions
import com.marpak.gmpolygon.InputPromptDialog.showInputDialog
import com.marpak.gmpolygon.databinding.ActivityMainBinding
import com.marpak.gmpolygon.enums.ButtonAction
import com.marpak.gmpolygon.listeners.BtnClickListener

class MainActivity : BaseActivity(), OnMapReadyCallback, OnMarkerDragListener, OnMapClickListener,
    BtnClickListener {
    private var googleMap: GoogleMap? = null
    private var mapClickListener: OnMapClickListener? = null

    private val polygonPoints: MutableList<LatLng> = ArrayList()
    private val crossLinePoints: MutableList<LatLng> = ArrayList()

    private var lastMarker: Marker? = null

    private var polyline: Polyline? = null
    private var polygon: Polygon? = null
    private var polygonOptions: PolygonOptions? = null

    private var binding: ActivityMainBinding? = null
    private var buttonPalette: ButtonPalette? = null

    private var btnListener: BtnClickListener? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        // Set the layout file as the content view.
        binding = ActivityMainBinding.inflate(layoutInflater)
        val view: View = binding!!.getRoot()
        setContentView(view)

        // Get a handle to the fragment and register the callback.
        val mapFragment =
            supportFragmentManager.findFragmentById(R.id.fragMap) as SupportMapFragment?
        mapFragment?.getMapAsync(this)
        btnListener = this

        buttonPalette = ButtonPalette(binding!!.buttonLayout, btnListener)
        setCustomTitle(text = "Map Activity")
    }

    // Get a handle to the GoogleMap object and display marker.
    override fun onMapReady(map: GoogleMap) {
        googleMap = map
        googleMap!!.setOnMarkerDragListener(this)

        mapClickListener = this
    }

    private fun addPoint(latLng: LatLng, pointsList: MutableList<LatLng>) {
        pointsList.add(latLng)
        if (polyline != null) {
            // remove existing polyline from map and add new polyline with updated points
            polyline!!.remove()
        }
        if (lastMarker == null) { // move a marker to last clicked point in the map
            lastMarker = googleMap!!.addMarker(
                MarkerOptions().position(latLng)
                    .icon(BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_AZURE))
            )
        } else {
            lastMarker!!.position = latLng
        }
        polyline = googleMap!!.addPolyline(
            PolylineOptions().addAll(pointsList)
                .color(getResources().getColor(R.color.polyline_color, null))
        )
    }

    private fun drawPolygon() {
        if (polygon != null) {
            // remove existing polygon from map and add new polygon with updated points
            polygon!!.remove()
        }
        polygonOptions = PolygonOptions().addAll(polygonPoints).strokeColor(
            getResources()
                .getColor(R.color.polyline_color, null)
        ).fillColor(getResources().getColor(R.color.polygon_fill_color, null))
        polygon = googleMap!!.addPolygon(polygonOptions!!)
        polygon!!.tag = "User Selection"
    }

    override fun onMarkerDrag(marker: Marker) {}
    override fun onMarkerDragEnd(marker: Marker) {
        polygonPoints[marker.title!!.toInt()] = marker.position
        drawPolygon()
    }

    override fun onMarkerDragStart(marker: Marker) {}
    override fun onMapClick(latLng: LatLng) {
        if (buttonPalette!!.currentAction == ButtonAction.START_DRAWING_SHAPE) {
            addPoint(latLng, polygonPoints)
        } else if (buttonPalette!!.currentAction == ButtonAction.START_DRAWING_CROSS_LINE) {
            addPoint(latLng, crossLinePoints)
        }
    }

    override fun onClicked(btn: Button) {
        when (btn) {
            binding!!.btnDrawPolygon -> {
                if (buttonPalette!!.currentAction == ButtonAction.CLEAR_DRAWINGS) {
                    buttonPalette!!.setAction(ButtonAction.START_DRAWING_SHAPE)

                    // start collecting click points
                    googleMap!!.setOnMapClickListener(mapClickListener)
                } else if (buttonPalette!!.currentAction == ButtonAction.START_DRAWING_SHAPE) {
                    // stop collecting click points and draw polygon
                    if (polygonPoints.size >= 3) {
                        drawPolygon()
                        buttonPalette!!.setAction(ButtonAction.STOP_DRAWING_SHAPE)
                        googleMap!!.setOnMapClickListener(null)
                    }
                }
            }

            binding!!.btnDrawCrossLine -> {
                if (buttonPalette!!.currentAction == ButtonAction.STOP_DRAWING_SHAPE) {
                    buttonPalette!!.setAction(ButtonAction.START_DRAWING_CROSS_LINE)

                    // start collecting click points
                    googleMap!!.setOnMapClickListener(mapClickListener)
                } else if (buttonPalette!!.currentAction == ButtonAction.START_DRAWING_CROSS_LINE) {
                    // stop collecting click points and draw cross line
                    if (polygonPoints.size >= 2) {
                        buttonPalette!!.setAction(ButtonAction.STOP_DRAWING_CROSS_LINE)
                        lastMarker!!.remove()
                        googleMap!!.setOnMapClickListener(null)
                    }
                }
            }

            binding!!.btnDivideCrossLine -> {
                // get user input via an prompt
                showInputDialog(this@MainActivity,
                    getResources().getString(R.string.prompt_dialog_title),
                    getResources().getString(R.string.prompt_dialog_description),
                    {
                        // divide the cross line
                        buttonPalette!!.setAction(ButtonAction.DIVIDE_CROSS_LINE)
                    }, InputPromptDialog.OnCancelListener {})
            }

            binding!!.btnResizePolygon -> {
                if (buttonPalette!!.currentAction == ButtonAction.DIVIDE_CROSS_LINE) {
                    // show markers on the corners of the polygon, let user drag those and resize it
                    for (index in polygonPoints.indices) {
                        val marker = googleMap!!.addMarker(
                            MarkerOptions()
                                .position(polygonPoints[index])
                                .icon(BitmapDescriptorFactory.defaultMarker(BitmapDescriptorFactory.HUE_AZURE))
                                .draggable(true)
                        )
                        if (marker != null) marker.title = index.toString()
                    }
                    buttonPalette!!.setAction(ButtonAction.START_RESIZING_SHAPE)
                } else if (buttonPalette!!.currentAction == ButtonAction.START_RESIZING_SHAPE) {
                    // hide markers on the map
                    googleMap!!.clear()
                    googleMap!!.addPolygon(polygonOptions!!)
                    buttonPalette!!.setAction(ButtonAction.STOP_RESIZING_SHAPE)
                }
            }

            binding!!.btnClearDrawings -> {
                lastMarker = null
                googleMap!!.clear()
                polygonPoints.clear()
                crossLinePoints.clear()
                buttonPalette!!.setAction(ButtonAction.CLEAR_DRAWINGS)
            }
        }
    }
}